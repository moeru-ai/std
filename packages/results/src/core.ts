export interface Err<E> {
  __type__: 'err'
  error: E
}

export interface None {
  __type__: 'none'
}

export interface Ok<T> {
  __type__: 'ok'
  value: T
}

export type Option<T> = None | Some<T>

export type Result<T, E> = Err<E> | Ok<T>

export interface Some<T> {
  __type__: 'some'
  value: T
}

export const err = <T, E>(error: E): Result<T, E> => ({
  __type__: 'err',
  error,
})

export const none: Option<never> = {
  __type__: 'none',
}

export const ok = <T, E>(value: T): Result<T, E> => ({
  __type__: 'ok',
  value,
})

export const some = <T>(value: T): Option<T> => ({
  __type__: 'some',
  value,
})
