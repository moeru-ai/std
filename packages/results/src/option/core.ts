export interface None {
  __type__: 'none'
}

export type Option<T> = None | Some<T>

export interface Some<T> {
  __type__: 'some'
  value: T
}

export const none: Option<never> = {
  __type__: 'none',
}

export const some = <T>(value: T): Option<T> => ({
  __type__: 'some',
  value,
})
