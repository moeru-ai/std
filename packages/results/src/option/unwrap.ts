import type { Option } from '../core'

import { isSome } from './is'

export const unwrap = <T>(o: Option<T>): T => {
  if (isSome(o))
    return o.value
  else
    throw new Error('called `option.unwrap()` on a `None` value')
}

export const unwrapOr = <T>(o: Option<T>, fallback: T): T =>
  isSome(o)
    ? o.value
    : fallback

export const unwrapOrElse = <T>(o: Option<T>, onNone: () => T): T =>
  isSome(o)
    ? o.value
    : onNone()

// eslint-disable-next-line sonarjs/no-identical-functions
export const unwrapOrElseAsync = async <T>(o: Option<T>, onNone: () => Promise<T>): Promise<T> =>
  isSome(o)
    ? o.value
    : onNone()
