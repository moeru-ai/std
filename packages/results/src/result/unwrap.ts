import type { Result } from './core'

import { isErr, isOk } from './is'

export const unwrap = <T, E>(r: Result<T, E>): T => {
  if (isOk(r))
    return r.value
  else if (r.error instanceof Error)
    throw r.error
  else
    throw new Error('called `result.unwrap()` on an `Err` value', { cause: r.error })
}

export const unwrapErr = <T, E>(r: Result<T, E>): E => {
  if (isErr(r))
    return r.error
  else
    throw new Error('called `result.unwrapErr()` on an `Ok` value', { cause: r.value })
}

export const unwrapOr = <T, E>(r: Result<T, E>, fallback: T): T =>
  isOk(r)
    ? r.value
    : fallback

export const unwrapOrElse = <T, E>(r: Result<T, E>, onErrValue: (e: E) => T): T =>
  isOk(r)
    ? r.value
    : onErrValue(r.error)

// eslint-disable-next-line sonarjs/no-identical-functions
export const unwrapOrElseAsync = async <T, E>(r: Result<T, E>, onErrValue: (e: E) => Promise<T>): Promise<T> =>
  isOk(r)
    ? r.value
    : onErrValue(r.error)
