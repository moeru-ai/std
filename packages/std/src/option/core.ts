export interface None {
  __mo: 'none'
}

export type Option<T> = None | Some<T>

export interface Some<T> {
  __mo: 'some'
  value: T
}

export const none: Option<never> = {
  __mo: 'none',
}

export const some = <T>(value: T): Option<T> => ({
  __mo: 'some',
  value,
})
