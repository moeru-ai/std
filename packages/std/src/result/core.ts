export interface Err<E> {
  __mr: 'err'
  error: E
}

export interface Ok<T> {
  __mr: 'ok'
  value: T
}

export type Result<T, E> = Err<E> | Ok<T>

export const err = <T, E>(error: E): Result<T, E> => ({
  __mr: 'err',
  error,
})

export const ok = <T, E>(value: T): Result<T, E> => ({
  __mr: 'ok',
  value,
})
