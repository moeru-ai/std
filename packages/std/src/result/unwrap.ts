import type { Result } from './core'

import { isOk, isErr } from './is'

export const unwrap = <T, E>(r: Result<T, E>): T => {
  if (isOk(r))
    return r.value
  else
    throw r.error
}

export const unwrapErr = <T, E>(r: Result<T, E>): E => {
  if (isErr(r))
    return r.error
  else
    throw new Error('called `result.unwrapErr()` on a `Ok` value', { cause: r.value })
}

export const unwrapOr = <T, E>(r: Result<T, E>, fallback: T): T =>
  isOk(r)
    ? r.value
    : fallback

export const unwrapOrElse = <T, E>(r: Result<T, E>, onErrValue: (e: E) => T): T =>
  isOk(r)
    ? r.value
    : onErrValue(r.error)

export const unwrapOrElseAsync = async <T, E>(r: Result<T, E>, onErrValue: (e: E) => Promise<T>): Promise<T> =>
  isOk(r)
    ? r.value
    : await onErrValue(r.error)
