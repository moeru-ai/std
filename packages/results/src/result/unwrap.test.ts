import { describe, expect, it } from 'vitest'

import { err, ok } from '../core'
import { unwrap, unwrapErr, unwrapOr, unwrapOrElse } from './unwrap'

describe('@moeru/results', () => {
  // https://doc.rust-lang.org/std/result/enum.Result.html#examples-19
  it('result.unwrap', () => {
    expect(unwrap(ok(2))).toBe(2)
    expect(() => unwrap(err('emergency failure'))).toThrowErrorMatchingSnapshot()
  })

  // https://doc.rust-lang.org/std/result/enum.Result.html#examples-22
  it('result.unwrapErr', () => {
    expect(() => unwrapErr(ok(2))).toThrowErrorMatchingSnapshot()
    expect(unwrapErr(err('emergency failure'))).toBe('emergency failure')
  })

  // https://doc.rust-lang.org/std/result/enum.Result.html#examples-29
  it('result.unwrapOr', () => {
    expect(unwrapOr(ok(9), 2)).toBe(9)
    expect(unwrapOr(err('error'), 2)).toBe(2)
  })

  // https://doc.rust-lang.org/std/result/enum.Result.html#examples-30
  it('result.unwrapOrElse', () => {
    let count = (x: string) => x.length

    expect(unwrapOrElse<number, string>(ok(2), count)).toBe(2)
    expect(unwrapOrElse(err('foo'), count)).toBe(3)
  })
})
