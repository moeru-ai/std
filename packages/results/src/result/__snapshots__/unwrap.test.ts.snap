// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@moeru/results > result.unwrap 1`] = `[Error: called \`result.unwrap()\` on an \`Err\` value]`;

exports[`@moeru/results > result.unwrapErr 1`] = `[Error: called \`result.unwrapErr()\` on an \`Ok\` value]`;
